diff --git a/CMakeLists.txt b/CMakeLists.txt
index b493984..a9b5b0a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -323,7 +323,7 @@ target_link_libraries(MEFISTO2 SMESHTrace TKernel)
 # Misc
 file(GLOB Misc_SRC src/Misc/*.cxx)
 add_library(SMESHMisc ${Misc_SRC})
-target_link_libraries(SMESHMisc SMESHTrace TKBRep TKShHealing TKMesh TKGeomAlgo TKTopAlgo TKG2d TKG3d TKV3d TKGeomBase TKBO)
+target_link_libraries(SMESHMisc SMESHTrace TKBRep TKShHealing TKMesh TKGeomAlgo TKTopAlgo TKG2d TKG3d TKGeomBase TKBO)
 if(WIN32)
     set_target_properties(SMESHMisc PROPERTIES COMPILE_FLAGS "-DUTILS_EXPORTS")
 endif()
