diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21cb116..a4eea62 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,6 +2,10 @@
 project(SMESH C CXX)
 cmake_minimum_required(VERSION 3.10)
 
+# setting the compile-features manually (needed by vtk) and not detected by cmake
+# for clan-compiler from anaconda
+SET(CMAKE_CXX_COMPILE_FEATURES "cxx_nullptr;cxx_override")
+
 
 # --------------------------------------------------------------------------- #
 # OPTIONS
