diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3e217c2..6001515 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,6 +2,10 @@
 cmake_minimum_required(VERSION 3.3)
 project(SMESH C CXX)
 
+# setting the compile-features manually (needed by vtk) and not detected by cmake
+# for clang-compiler from anaconda
+SET(CMAKE_CXX_COMPILE_FEATURES "cxx_nullptr;cxx_override")
+
 # --------------------------------------------------------------------------- #
 # OPTIONS
 # --------------------------------------------------------------------------- #
